src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -180,6 +191,17 @@ public JsonConverter GetConverter(Type typeToConvert)
             // Priority 4: Attempt to get built-in converter.
             if (converter == null)
             {
+                if (s_defaultSimpleConverters == null || s_defaultFactoryConverters == null)
+                {
+                    // (De)serialization using serializer's options-based methods has not yet occurred, so the built-in converters are not rooted.
+                    // Even though source-gen code paths do not call this method <i.e. JsonSerializerOptions.GetConverter(Type)>, we do not root all the
+                    // built-in converters here since we fetch converters for any type included for source generation from the binded context (Priority 1).
+                    Debug.Assert(s_defaultSimpleConverters == null);
+                    Debug.Assert(s_defaultFactoryConverters == null);
+                    ThrowHelper.ThrowNotSupportedException_BuiltInConvertersNotRooted(typeToConvert);
```

Comment:
I'm not sure I understand what use case could trigger this particular failure mode. What actions should the user be taking to resolve it? The corresponding [error message](https://github.com/dotnet/runtime/pull/51300/files#diff-f343f56e41ad406150ed9d35e7548f0d118dd3038c57dff2ff309e5e96331091R606) isn't very clear on that.

If this code path could only be entered because of a bug in STJ then this should be a plain assert.

Reply:
This will be hit when a project creates an options instance and calls `JsonSerializerOptions.GetConverter(Type)` without first performing (de)serialization with an `options`-based overload of the serializer. The built-in types are not rooted at this point, hence the exception. This is a breaking change since we go from returning a converter for almost any type, to throwing `NotSupportedException`.

The mitigation is as as follows:

- Indicate all serializable types, or types you wish to fetch a converter for with `JsonSerializableAttribute`
- Register the generated `JsonContext` class with your options instance:

    ```cs
    options.AddContext<JsonContext>();
    ```
- `options.GetConverter(Type)` works again.
    
The approach now is to assume that if the converters are not rooted, then the app authors care about size and are using the code-gen model described above. The source generator itself does not generate calls to `Options.GetConverter`,  so we can go either way here (root the converters for backward-compat, or not root them i.e. depend on source generation).

---

For now, I've updated exception message to include more info about how to mitigate it. Eventually we'll want source-gen related exception messages to point to a central docs.microsoft.com page that can expound as necessary.